### PR TITLE
Use JCTools MpscLinkedAtomicUnpaddedQueue as FiberRuntime inbox (fixes #8807)

### PR DESCRIPTION
**Pull request by drewpcodes**
TL;DR; Improved above 5% in 11 of 23 tested benchmarks, topping out at +28% in zioDeepAsyncMany

After careful testing, it appears that the referenced collection provides similar performance characteristics to Java's ConcurrentLinkedQueue but with better throughput under load. Many alternatives were compared: from simple synchronous blocks around standard collections, to circular buffers, and various array-based concurrent queues from JCTools.

While in the general case, array-based concurrent queues such as those modeled after fast-flow should provide the best overall performance: we do not fit the general use case with the FiberRuntime.

1. The FiberRuntime is sensitive to allocations; even with a small fixed buffer size (4) the extra allocation cost more throughput than it made up for with cache-friendliness.
2. The additional allocations for linked queue items are "lazier" than the hit taken during initialization with array-based queues.
3. Specifically regarding false-sharing, the FiberRuntime is not as sensitive to false-sharing as may be expected and the *Unpadded variants of the JCTools collections win out again based on reduced memory allocation.

To note: while the CAS operations used by AtomicReference under the hood of this collection are quite fast, there have been breakthroughts in the past few years around bounded and unbounded queues based on the XAdd atomic instructions which are notably faster than CAS under load. This is an avenue that may be worth exploring, but the excellent performance of the JCTools collections coupled with their code quality makes for a compelling reason to not "roll our own" unless _very significant_ gains are still possible. The FiberRuntime has other performance areas of concern.

I will add detailed benchmark output in another comment, but here are the highlights from ZIO's benchmark suite:
image

/claim #8807

### CODE REVIEW

**Comment by drewpcodes** [comment]
In addition to the benchmark suite in ZIO (`benchmarks/Jmh/run ^(?!.*(akka|cats|chunks|cio|completableFuture|fs2|future|mono|rx|stm|twitter|ConfigPro|CustomExe|FiberRef|Hub|NewEnc|Queue|Sort|Stream|ZEnv).)(.*(Async|FlatMap|Fork|Race)) -f 5`), I also grabbed the [TechEmpower zio-http benchmark](https://github.com/TechEmpower/FrameworkBenchmarks/compare/master...drewpcodes:FrameworkBenchmarks:master) and updated it to the latest zio-http. I then ran it with the current 2.x version of zio, then compared it with the code in this PR. The synthetic benchmark results seem to transfer well to zio-http benchmarks.

It's easy to note the increase in requests/sec, reduced average response times, and general upward trend in success rate (with some random variability considering this was done on my desktop).
### zio 2.1.1
#### plaintext
```shell
> oha -n 50000 -c 5000 --latency-correction --disable-keepalive http://localhost:8080/plaintext
Summary:
  Success rate: 97.95%
  Total:        6.9676 secs
  Slowest:      2.2577 secs
  Fastest:      0.0094 secs
  Average:      0.5431 secs
  Requests/sec: 7176.0724

  Total data:   621.75 KiB
  Size/request: 13 B
  Size/sec:     89.23 KiB

Response time histogram:
  0.009 [1]     |
  0.234 [18674] |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.459 [6338]  |■■■■■■■■■■
  0.684 [10473] |■■■■■■■■■■■■■■■■■
  0.909 [2375]  |■■■■
  1.134 [5515]  |■■■■■■■■■
  1.358 [1380]  |■■
  1.583 [640]   |■
  1.808 [2217]  |■■■
  2.033 [0]     |
  2.258 [1362]  |■■

Response time distribution:
  10.00% in 0.0332 secs
  25.00% in 0.0923 secs
  50.00% in 0.4080 secs
  75.00% in 0.7729 secs
  90.00% in 1.2002 secs
  95.00% in 1.6285 secs
  99.00% in 2.1554 secs
  99.90% in 2.2136 secs
  99.99% in 2.2555 secs


Details (average, fastest, slowest):
  DNS+dialup:   0.4791 secs, 0.0002 secs, 2.1781 secs
  DNS-lookup:   0.0000 secs, 0.0000 secs, 0.0069 secs

Status code distribution:
  [200] 48975 responses
```
#### json
```shell
> oha -n 50000 -c 5000 --latency-correction --disable-keepalive http://localhost:8080/json
Summary:
  Success rate: 99.30%
  Total:        6.7902 secs
  Slowest:      2.4814 secs
  Fastest:      0.0141 secs
  Average:      0.5606 secs
  Requests/sec: 7363.5341

  Total data:   1.28 MiB
  Size/request: 27 B
  Size/sec:     192.79 KiB

Response time histogram:
  0.014 [1]     |
  0.261 [17858] |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.508 [8898]  |■■■■■■■■■■■■■■■
  0.754 [10690] |■■■■■■■■■■■■■■■■■■■
  1.001 [2759]  |■■■■
  1.248 [4338]  |■■■■■■■
  1.494 [1442]  |■■
  1.741 [2018]  |■■■
  1.988 [676]   |■
  2.235 [593]   |■
  2.481 [376]   |

Response time distribution:
  10.00% in 0.0583 secs
  25.00% in 0.1740 secs
  50.00% in 0.4031 secs
  75.00% in 0.7466 secs
  90.00% in 1.2583 secs
  95.00% in 1.6563 secs
  99.00% in 2.1887 secs
  99.90% in 2.3583 secs
  99.99% in 2.4007 secs


Details (average, fastest, slowest):
  DNS+dialup:   0.4694 secs, 0.0003 secs, 2.3671 secs
  DNS-lookup:   0.0000 secs, 0.0000 secs, 0.0080 secs

Status code distribution:
  [200] 49649 responses
```
### New inbox
#### plaintext
```shell
> oha -n 50000 -c 5000 --latency-correction --disable-keepalive http://localhost:8080/plaintext
Summary:
  Success rate: 98.14%
  Total:        6.2450 secs
  Slowest:      2.3922 secs
  Fastest:      0.0005 secs
  Average:      0.5240 secs
  Requests/sec: 8006.4388

  Total data:   622.97 KiB
  Size/request: 13 B
  Size/sec:     99.75 KiB

Response time histogram:
  0.001 [1]     |
  0.240 [18532] |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.479 [8125]  |■■■■■■■■■■■■■■
  0.718 [9907]  |■■■■■■■■■■■■■■■■■
  0.957 [3654]  |■■■■■■
  1.196 [3404]  |■■■■■
  1.436 [2953]  |■■■■■
  1.675 [657]   |■
  1.914 [1353]  |■■
  2.153 [223]   |
  2.392 [262]   |

Response time distribution:
  10.00% in 0.0490 secs
  25.00% in 0.1439 secs
  50.00% in 0.3519 secs
  75.00% in 0.7270 secs
  90.00% in 1.2200 secs
  95.00% in 1.5850 secs
  99.00% in 1.9068 secs
  99.90% in 2.3250 secs
  99.99% in 2.3899 secs


Details (average, fastest, slowest):
  DNS+dialup:   0.4409 secs, 0.0002 secs, 2.3575 secs
  DNS-lookup:   0.0000 secs, 0.0000 secs, 0.0096 secs

Status code distribution:
  [200] 49071 responses
```
#### json
```shell
> oha -n 50000 -c 5000 --latency-correction --disable-keepalive http://localhost:8080/json
Summary:
  Success rate: 98.00%
  Total:        6.5426 secs
  Slowest:      2.3449 secs
  Fastest:      0.0007 secs
  Average:      0.5164 secs
  Requests/sec: 7642.1953

  Total data:   1.26 MiB
  Size/request: 27 B
  Size/sec:     197.46 KiB

Response time histogram:
  0.001 [1]     |
  0.235 [22008] |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.470 [3995]  |■■■■■
  0.704 [10589] |■■■■■■■■■■■■■■■
  0.938 [2056]  |■■
  1.173 [4982]  |■■■■■■■
  1.407 [1697]  |■■
  1.642 [1192]  |■
  1.876 [1505]  |■■
  2.110 [199]   |
  2.345 [774]   |■

Response time distribution:
  10.00% in 0.0322 secs
  25.00% in 0.0757 secs
  50.00% in 0.3296 secs
  75.00% in 0.7085 secs
  90.00% in 1.1911 secs
  95.00% in 1.6426 secs
  99.00% in 2.1879 secs
  99.90% in 2.2835 secs
  99.99% in 2.3354 secs


Details (average, fastest, slowest):
  DNS+dialup:   0.4552 secs, 0.0002 secs, 2.2692 secs
  DNS-lookup:   0.0000 secs, 0.0000 secs, 0.0106 secs

Status code distribution:
  [200] 48998 responses
```

**Comment by kyri-petrou** [comment]
Hey @drewpcodes, there are a few things to unpack here so I'll try my best to do it in a single comment:

> I then ran it with the current 2.x version of zio, then compared it with the code in this PR

We recently merged in some improvements to the FiberRuntime that have not yet been released (most notably, [this PR](https://github.com/zio/zio/pull/9016)). Could you compare against the latest snapshot to make sure that we're not including improvements from other PRs?

> This is an avenue that may be worth exploring, but the excellent performance of the JCTools collections coupled with their code quality makes for a compelling reason to not "roll our own" unless very significant gains are still possible

Look, personally I agree. At first I was _really_ sceptical about this, but after looking at the jctools-core suite, I'm actually warming up to the idea because it contains a lot of datastructures that ZIO could benefit from, and help us reduce code maintenance burden. However, if we're going to go down this avenue, we still need to keep the "zero-dependency contract" that ZIO has. To do this, we would need to shade the `jctools-core` dependency. We also need to make it cross-compilable across JVM/JS/Native, which I'm guessing this PR doesn't currently account for.

I'll leave some more comments in the code

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala [hunk 1]
```diff
@@ -244,7 +244,7 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
 
       while (evaluationSignal == EvaluationSignal.Continue) {
         evaluationSignal = {
-          val message = inbox.poll()
+          val message = inbox.relaxedPoll()
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/internal/FiberRuntime.scala]: I'm not 1000% sold about using `relaxedPoll` here. Which benchmarks did you notice an improvement for by using `relaxedPoll` here?

**Comment by kyri-petrou** [comment]
Also, can you please rebase off the current series/2.x branch and rerun the benchmarks? Just wanna make sure that the changes we did recently are not affecting the benchmark results in either way

**Comment by drewpcodes** [comment]
@kyri-petrou Thanks for taking a look so quickly! I fully expected some back and forth and tweaks here so I'm happy to get that underway.

> Could you compare against the latest snapshot to make sure that we're not including improvements from other PRs?

I will re-run my benchmarks and dump the results in here, including results for: the latest released snapshot, the head of the series/2.x branch, and my branch rebased off series/2.x. Let me know if there are other combinations you would like to compare.

**I've updated the description for the zio-http benchmarks, as they were actually run against zio 2.1.1 off maven and compared to my branch**

> However, if we're going to go down this avenue, we still need to keep the "zero-dependency contract" that ZIO has

Happy to explore this with you; I was fully expecting to need to remove it as a dependency and do something like shade it or bring the minimum source into the zio tree, and I should have led with that.

> We also need to make it cross-compilable across JVM/JS/Native

Yes, I have not attempted that yet. That will give some guidance in how to approach the jctools dependency and I will run some builds and see what happens.

**Would you mind pointing me in the direction of some notes/docs on scala.js and fibers?** Based on the existing code I had assumed scala.js did not have support for this particular `FiberRuntime` (since it was already making use of JDK classes for things like the inbox).

**Comment by kyri-petrou** [comment]
> I will re-run my benchmarks and dump the results in here, including results for: the latest released snapshot, the head of the series/2.x branch, and my branch rebased off series/2.x

You should choose just one of the latest released snapshot or the head of the series/2.x branch. We publish a new snapshot whenever we merge to series/2.x so they should be more or less identical.

Oh, if you're benchmarking code compiled from source, make sure to set the `CI_RELEASE_MODE=1` when compiling/generating the artifacts to enable some additional optimizations which are normally enabled only when we release snapshot / stable versions.

> Would you mind pointing me in the direction of some notes/docs on scala.js and fibers

I don't think there is much to point to. What we do is that we place JVM-specific code under the `/core/jvm` directory, and then use something that's available for JS / Native and put it under the `/js` and `/native` directories. For js / native you can just use `java.util.concurrent.ConcurrentLinkedQueue` since that one is available for those.

If you take a look at the linked issue, you'll see that there was a previous attempt at this ticket which did something similar

**Comment by drewpcodes** [comment]
@kyri-petrou Here is a table outlining the improved benchmarks when comparing #8168cd to this PR, with CI_RELEASE_MODE set.
image

There is one statistically significant decrease in performance, and it's actually quite a big decrease, and that's in the `zioFixedThreadPoolForkMany` test. That appears to have ~30% decrease in performance. I'm not sure what the root cause is yet, but it appears to be consistent between runs so it shouldn't be difficult to idenitfy. Any other tests that decrease in performance tend to go <=1% one way or the other.

**Comment by kyri-petrou** [comment]
I think I know what it might be. Can you change the poll in drainQueueAfterAsync to a strict one and rerun them?

Also try changing the top-level one and see if that makes a difference aa well

**Comment by drewpcodes** [comment]
@kyri-petrou so I ran some performance testing on just that scheduler benchmark and found something interesting. It doesn't appear that the new mailbox has anything to do with the performance issue there (I'm observing this behavior on the head of series/2.x as well). In fact, in JMH warmup or iterations I can observe >100% swings in op/s in the same JVM fork.

I haven't been able to observe the behavior under a profiler, which I think may be due to the overhead creating more fairness between threads. The blocking linked queue in the scheduler is top of the heat map, so I'm wondering if some threads in the pool are being repeatedly blocked waiting on that collection.

**Comment by kyri-petrou** [comment]
Yeah this is definitely weird. I think this has to do a fair bit with the fact that in that benchmark we're not using the `ZScheduler` (which is highly optimized for usage with ZIO's fibers). What happens if you use the virtual threads-based pool for that benchmark?

**Comment by kyri-petrou** [comment]
@drewpcodes I came across something quite interesting yesterday. It seems that in Netty they have a boolean flag which is determined at runtime whether the platform supports `sun.misc.Unsafe` and uses implementations that are more performant when it does. Maybe we should also implement something similar and use the `org.jctools.queues.unpadded.MpscLinkedUnpaddedQueue` queue instead when it does?

This is where / how it's [implemented in Netty](https://github.com/netty/netty/blob/20a790ed362a3c11e0e990b58598e4ac6aa88bef/common/src/main/java/io/netty/util/internal/PlatformDependent.java#L376-L378). We could add it along the other platform-specific flags in [this file](https://github.com/zio/zio/blob/16e72493508a97ffec3790f7e84c78c77b4d94d1/core/jvm/src/main/scala/zio/internal/PlatformSpecific.scala#L84-L85)

**Comment by drewpcodes** [comment]
@kyri-petrou It appears that the JCTools implementation already supports fallbacks when unsafe is unavailable (it checks internally).

I don't know what level of involvement you have with the zio org, so this may not be a question you can answer: do you happen to know if the benchmark performance we see in the above examples is enough to justify shading JCTools and bringing it in as a JVM dependency?

If we can answer that question with 'yes', then I will proceed with that and swap the unsupported platform(s) like Scala.js over to the original concurrent queue. If not, then I'm going to try and implement the same algorithm without the use of JCTools.

I'm personally ok with either outcome, I had originally thought simply using JCTools would be an easy win but if there's concern about the long-term decision to bring in an external dependency (even if shaded) then it might be best to reimplement knowing what the benchmarks will look like.

**Comment by jdegoes** [comment]
I think shading can make sense because it reduces maintenance burden in this project. 👍

File: core/shared/src/main/scala/zio/internal/FiberRuntime.scala (hunk 1 above)
Comment (drewpcodes) [reply core/shared/src/main/scala/zio/internal/FiberRuntime.scala]: In my latest iteration I have removed the use of relaxedPoll because it seems to only benefit synthetic workloads. In real-world usage it seems the use of the `spinWait` likely puts us ahead in the small window where the queue may not be empty but the message is not yet visible to the consumer.

**Comment by drewpcodes** [comment]
@kyri-petrou @jdegoes 
Thank you both. I went ahead and pushed the changes to the JVM build settings to shade jctools, and cleaned things up so that js and native build again using the previous queue impl.

File: core/jvm/src/main/scala/zio/internal/FiberInbox.scala [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package zio.internal
+
+import org.jctools.queues.atomic.unpadded.MpscLinkedAtomicUnpaddedQueue
+
+private[zio] final class FiberInbox extends Serializable {
+  private val inbox = new MpscLinkedAtomicUnpaddedQueue[FiberMessage]()
+
+  def poll(): FiberMessage = inbox.poll()
```
Comment (kyri-petrou) [inline core/jvm/src/main/scala/zio/internal/FiberInbox.scala]: Don't we also need the `relaxed*` methods here to get the benefit of those wherever we need them?

File: build.sbt [hunk 3]
```diff
@@ -214,7 +214,11 @@ lazy val core = crossProject(JSPlatform, JVMPlatform, NativePlatform)
   .settings(scalacOptions += "-Wconf:msg=[zio.stacktracer.TracingImplicits.disableAutoTrace]:silent")
   .jvmSettings(
     replSettings,
-    mimaSettings(failOnProblem = true)
+    mimaSettings(failOnProblem = true),
+    libraryDependencies += "org.jctools" % "jctools-core" % "4.0.5",
+    assemblyShadeRules += ShadeRule
+      .rename("org.jctools.core.**" -> "zio.shade.@0")
+      .inAll // Shaded as zio.shade.org.jctools.core.foo
```
Comment (kyri-petrou) [inline build.sbt:221]: I've never seen this before, does it work? Isn't this just when we create uber jars with the assembly plugin?

I think we need to use [this plugin](https://github.com/coursier/sbt-shading) for shading instead

File: core/jvm/src/main/scala/zio/internal/FiberInbox.scala [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package zio.internal
+
+import org.jctools.queues.atomic.unpadded.MpscLinkedAtomicUnpaddedQueue
```
Comment (kyri-petrou) [inline core/jvm/src/main/scala/zio/internal/FiberInbox.scala:3]: Shouldn't we use `import org.jctools.queues.unpadded.MpscLinkedUnpaddedQueue` instead?

**Review by kyri-petrou — COMMENTED** [review]
I see that you removed usages of `relaxedPoll`, shouldn't we keep them? Isn't those that were offering us the performance improvement?

File: core/jvm/src/main/scala/zio/internal/FiberInbox.scala (hunk 2 above)
Comment (drewpcodes) [reply core/jvm/src/main/scala/zio/internal/FiberInbox.scala]: I've revisited that again, and after running some benchmarks overnight it appears there are no statistically significant costs that I can measure when using `relaxedPoll` under high contention. This is because we do a final check on the inbox `isEmpty` before giving up, which catches the case where `relaxedPoll` is unable to see the latest item offered during high contention. 

In terms of benefit, I am able to measure a consistent 6-7% throughput improvement in the `zioDeepAsyncOne` benchmark (as an easy example) due to the consistent optimistic polling wins exhibited therein.

So in my latest commit I've enabled relaxed polling under the JVM.

File: core/jvm/src/main/scala/zio/internal/FiberInbox.scala (hunk 4 above)
Comment (drewpcodes) [reply core/jvm/src/main/scala/zio/internal/FiberInbox.scala:3]: So their class hierarchy can look a bit messy due to their codegen, but the difference is the `atomic` variant of the queue I've chosen uses atomic references rather than memory barriers for its nodes which performs better under all scenarios in our benchmarks.

File: build.sbt (hunk 3 above)
Comment (drewpcodes) [reply build.sbt:221]: That is correct, it shades the dependency only in that uber jar scenario. I had assumed we were only publishing uber jars but I see that not all targets set publish/skip to true.

It seems like [we may need both then](https://github.com/coursier/sbt-shading/issues/13)? What do you think?

Comment (kyri-petrou) [reply build.sbt:221]: Can you check whether using `relaxedPoll` in `drainQueueWhileRunning` method makes a difference in the `BroadFlatMapBenchmark` and `NarrowFlatMapBenchmark` cases? That's where I expect them to make the most difference

File: core/jvm/src/main/scala/zio/internal/FiberInbox.scala (hunk 4 above)
Comment (kyri-petrou) [reply core/jvm/src/main/scala/zio/internal/FiberInbox.scala:3]: What is the difference you're seeing in the benchmarks and in which ones?

File: core/jvm/src/main/scala/zio/internal/FiberInbox.scala (hunk 2 above)
Comment (kyri-petrou) [reply core/jvm/src/main/scala/zio/internal/FiberInbox.scala]: Can you check whether using `relaxedPoll` in `drainQueueWhileRunning` method makes a difference in the `BroadFlatMapBenchmark` and `NarrowFlatMapBenchmark` cases? That's where I expect them to make the most difference

File: core/jvm/src/main/scala/zio/internal/FiberInbox.scala (hunk 4 above)
Comment (drewpcodes) [reply core/jvm/src/main/scala/zio/internal/FiberInbox.scala:3]: @kyri-petrou it takes about 2 hours to run the benchmark suite each time, so 4 hours to do a mostly fair comparison (I have an idle machine I'm using to run the benchmarks to avoid as much inconsistency as possible, but I try to still run them close together).

One that very clearly demonstrated the difference between atomics and barriers is `DeepAsyncBenchmark.zioDeepAsyncOne`, I can get you a fresh run of that quite quickly and I'll reply with the numbers in this thread.

If there are others that you specifically would like to compare atomic vs barrier please let me know.

File: build.sbt (hunk 3 above)
Comment (kyri-petrou) [reply build.sbt:221]: I think we need both yes

File: core/jvm/src/main/scala/zio/internal/FiberInbox.scala (hunk 4 above)
Comment (kyri-petrou) [reply core/jvm/src/main/scala/zio/internal/FiberInbox.scala:3]: That's OK, I think we can keep the atomic-based one for the time being and revisit this in the future if we want to

File: core/jvm/src/main/scala/zio/internal/FiberInbox.scala [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+private[zio] final class FiberInbox extends Serializable {
+  private val inbox = new MpscLinkedAtomicUnpaddedQueue[FiberMessage]()
+
+  def poll(): FiberMessage = inbox.relaxedPoll()
```
Comment (kyri-petrou) [inline core/jvm/src/main/scala/zio/internal/FiberInbox.scala:8]: I've seen you changed this to use `relaxedPoll`. We definitely don't want that. We need to be selective on where to use it, and _I think_ the only place that it really makes sense is in the `drainQueueWhileRunning` method. Also see my comment [here](https://github.com/zio/zio/pull/9036#discussion_r1701422436)

**Comment by kyri-petrou** [comment]
@drewpcodes are you still planning on working on this? AFAICT we only need shading and to finalize which variants of `poll` we'll use in different methods of `FiberRuntime`

**Comment by kyri-petrou** [comment]
@jdegoes it seems that this PR has gone stale, but I think the approach is the right one (once [these are sorted out](https://github.com/zio/zio/pull/9036#issuecomment-2285767302)). What's the best approach here? Close the PR and open a new one?

Also pinging @drewpcodes in case you're still interested on working on this

**Comment by kyri-petrou** [comment]
Hey @drewpcodes, this PR has been stale for more than 2 months now and haven't heard from your since so I'll close it.

Since I'd really like to add support for JUC queues and other structures in ZIO, I'll open a PR to add it as a shaded dependency and implement the fiber mailbox along with adding results for the relevant benchmarks. If you're still around I'm happy to share the bounty with you; just let me know.

**Comment by HollandDM** [comment]
what is the status of this guys, do we still plan on shading jctools into the codebase?
